core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -267,15 +270,21 @@ public boolean isVersionActual(int segment, EntryVersion version) {
    }
 
    @Override
-   public void notifyKeyTransferFinished(int segment, boolean expectValues) {
-      SegmentState update = expectValues ? SegmentState.VALUE_TRANSFER : SegmentState.OWNED;
-      SegmentState previous;
-      do {
-         previous = segmentStates.get(segment);
-         // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
-         // when the CONFIRM_REVOKED_SEGMENTS failed.
-         log.tracef("Finishing transfer for segment %d = %s -> %s", segment, previous, update);
-      } while (!segmentStates.compareAndSet(segment, previous, update));
+   public void notifyKeyTransferFinished(int segment, boolean expectValues, boolean cancelled) {
+      SegmentState update;
+      if (expectValues) {
+         update = SegmentState.VALUE_TRANSFER;
+      } else if (cancelled) {
+         // The transfer is cancelled when a newer topology is being installed.
+         update = SegmentState.NOT_OWNED;
+      } else {
+         update = SegmentState.OWNED;
+      }
+      // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
+      // when the CONFIRM_REVOKED_SEGMENTS failed.
+      // Another possibility is when there is already another rebalance starting, then it might be NOT_OWNED
```

Comment:
If I understand the problem correctly, our solution for dist/repl caches is to use one `LimitedExecutor` per cache so that the new rebalance cannot start until the old topology update was processed and the old inbound transfers were all cancelled. The new coord always sends a topology update before sending the rebalance, so it doesn't matter if it was 2 crashes in a row, crash + join, or join + crash. The only complication is that the topology update and the rebalance start commands are not ordered, so we need to create the fake topology update in `LTMI.resetLocalTopologyBeforeRebalance`.

Now `ScatteredStateConsumerImpl.beforeTopologyInstalled` also cancels old inbound transfers, so I thought you won't have unfinished transfers from a previous rebalance any more.